### PR TITLE
Add HKDF expander interface

### PR DESCRIPTION
**Pull request by Ryanfsdf**
Added the HKDF expander interface. This interface should allow us to use various crypto libraries to implement the HKDF expand label function specified in Section 7 of https://tools.ietf.org/html/rfc8446. A followup to this PR will include an implementation using Go's crypto libraries.

### CODE REVIEW

File: security/s2a/internal/conn/expander.go [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package conn
```
Comment (matthewstevenson88) [inline security/s2a/internal/conn/expander.go:1]: Please put this file in `security/s2a/internal/crypter`. (All of the code for our record protocol will go in this directory, rather than a `conn` directory like ALTS has.)

File: security/s2a/internal/conn/expander.go [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+
+import "hash"
+
+// HKDFExpander is the interface for an HMAC-based Extract-and-Expand Key
+// Derivation Function (HKDF) expander. This is specified in Section 7 of
+// https://tools.ietf.org/html/rfc8446
+type HKDFExpander interface {
+	ExpandLabel(hash func() hash.Hash, secret, label []byte, length int) ([]byte, error)
```
Comment (matthewstevenson88) [inline security/s2a/internal/conn/expander.go:9]: Can we call the method `Expand`?

File: security/s2a/internal/conn/expander.go [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package conn
+
+import "hash"
+
+// HKDFExpander is the interface for an HMAC-based Extract-and-Expand Key
+// Derivation Function (HKDF) expander. This is specified in Section 7 of
```
Comment (matthewstevenson88) [inline security/s2a/internal/conn/expander.go:6]: Can you change `Section 7` to `Section 7.2`? And make the link to that section instead? I.e.

https://tools.ietf.org/html/rfc8446#section-7.2

Comment (matthewstevenson88) [reply security/s2a/internal/conn/expander.go:6]: Also, can you change the start of the second sentence to `It's use in TLS 1.3 is specified in...`?

File: security/s2a/internal/conn/expander.go [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package conn
+
+import "hash"
+
+// HKDFExpander is the interface for an HMAC-based Extract-and-Expand Key
```
Comment (matthewstevenson88) [inline security/s2a/internal/conn/expander.go:5]: Please replace `an HMAC-based Extract-and-Expand Key Derivation Function (HKDF) expander` with `the HKDF expansion function; see https://tools.ietf.org/html/rfc5869 for details.

(We want to make it clear that we are using **the** HKDF function as specified in RFC 5869, as opposed to some HMAC-based Extract-and-Expand key derivation function, which could be ambiguous.)

**Comment by matthewstevenson88** [comment]
Thanks for the PR Ryan! I've left a few minor comments.

File: security/s2a/internal/conn/expander.go (hunk 2 above)
Comment (Ryanfsdf) [reply security/s2a/internal/conn/expander.go:9]: I think that may lead to a bit of confusion since the RFC makes a distinction between Expand and Expand Label. WDYT?

Comment (matthewstevenson88) [reply security/s2a/internal/conn/expander.go:9]: The "HKDF-Expand-Label" terminology is, as far as I'm aware, local to the TLS 1.3 RFC. Indeed, they define it as the start of section 7 in the terms of "HKDF-Expand". The latter function, "HKDF-Expand", is the more commonly used and known (and it even has its own RFC). For that reason, I think it's better to use "HKDF-Expand". WDYT?

Comment (Ryanfsdf) [reply security/s2a/internal/conn/expander.go:9]: Changed as per discussion.

File: security/s2a/internal/conn/expander.go (hunk 1 above)
Comment (Ryanfsdf) [reply security/s2a/internal/conn/expander.go:1]: Done

File: security/s2a/internal/conn/expander.go (hunk 3 above)
Comment (Ryanfsdf) [reply security/s2a/internal/conn/expander.go:6]: Done

File: security/s2a/internal/conn/expander.go (hunk 4 above)
Comment (Ryanfsdf) [reply security/s2a/internal/conn/expander.go:5]: Done

**Comment by Ryanfsdf** [comment]
Addressed the review comments, PTAL!

**Comment by matthewstevenson88** [comment]
Thanks for the PR Ryan! Added Cesar for Go readability.

File: security/s2a/internal/crypter/expander.go [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package crypter
+
+import "hash"
+
+// HKDFExpander is the interface for the HKDF expansion function; see
+// https://tools.ietf.org/html/rfc5869 for details. It's use in TLS 1.3 is specified in
+// https://tools.ietf.org/html/rfc8446#section-7.2
+type HKDFExpander interface {
```
Comment (cesarghali) [inline security/s2a/internal/crypter/expander.go:8]: Why do we need an interface here and can `Expand` be an unexported function in the package where it will be used?

Comment (Ryanfsdf) [reply security/s2a/internal/crypter/expander.go:8]: I think we'll mostly likely be using it in the `crypter` package only, since I think we're placing everything related to the S2A crypter in this package. I'll unexport both `HKDFExpander` as well as the `Expand`.

As for why we need an interface, @matthewstevenson88 had some ideas about why an interface here may be useful. We can discuss this during standup.

Comment (matthewstevenson88) [reply security/s2a/internal/crypter/expander.go:8]: We have an interface here in case we need to link in a different crypto library in the future. This way we would need to change any of the logic that depends on HKDF-expand. We can discuss more during the standup?

Comment (cesarghali) [reply security/s2a/internal/crypter/expander.go]: Thanks both for the clarification. Sorry I missed today's standup but let's discuss in tomorrow's.

File: security/s2a/internal/crypter/expander.go [hunk 6]
```diff
@@ -0,0 +1,12 @@
+package crypter
+
+import (
+	"hash"
+)
+
+// hkdfExpander is the interface for the HKDF expansion function; see
+// https://tools.ietf.org/html/rfc5869 for details. It's use in TLS 1.3 is specified in
+// https://tools.ietf.org/html/rfc8446#section-7.2
+type hkdfExpander interface {
```
Comment (cesarghali) [inline security/s2a/internal/crypter/expander.go:10]: If we're unexporting this, then we should also unexport `S2AAeadCrypter`. We can do that in a separate PR after #8 is submitted. WDYT?

File: security/s2a/internal/crypter/expander.go [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+	expand(hash func() hash.Hash, secret, label []byte, length int) ([]byte, error)
```
Comment (cesarghali) [inline security/s2a/internal/crypter/expander.go:11]: Please add a comment about what this function does, specifically what are the inputs and outputs.

File: security/s2a/internal/crypter/expander.go (hunk 6 above)
Comment (Ryanfsdf) [reply security/s2a/internal/crypter/expander.go:10]: Sounds good. When I added `S2AAeadCrypter` I didn't realize that it would only be used in the `crypter` package. I'll unexport it in a separate PR.

File: security/s2a/internal/crypter/expander.go (hunk 7 above)
Comment (Ryanfsdf) [reply security/s2a/internal/crypter/expander.go:11]: Done
